### PR TITLE
[TACHYON-791] Add permission model

### PR DESCRIPTION
**Pull request by shenguoquan**
https://tachyon.atlassian.net/browse/TACHYON-791
Tachyon is an inode based filesystem, an inode present a file or a directory. So the permission model is put in the inode structure. The Google doc(https://docs.google.com/document/d/1anFhvJpajX3EtZFatHQ9Yz3-PWh3Nv6OvcLrMjDID_Q) describes the permision in detail [section 3.1]

### CODE REVIEW

**Comment by shenguoquan** [comment]
@apc999 @ooq , Please help me to review this PR. Thanks

File: common/src/main/java/tachyon/Constants.java [hunk 1]
```diff
@@ -248,4 +248,7 @@
   public static final String TACHYON_SECURITY_AUTHENTICATION = "tachyon.security.authentication";
   public static final String TACHYON_AUTHENTICATION_PROVIDER_CUSTOM_CLASS =
       "tachyon.authentication.provider.custom.class";
+  // Authorization
+  public static final String FS_PERMISSIONS_UMASK_KEY = "tfs.permission.umask";
```
Comment (apc999) [inline common/src/main/java/tachyon/Constants.java]: Can we have this property string consistent with other Tachyon properties, maybe something like
"tachyon.fs.permission.umask"?

File: common/src/main/java/tachyon/security/authorization/FsAction.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+  EXECUTE("--x"),
+  WRITE("-w-"),
+  WRITE_EXECUTE("-wx"),
+  READ("r--"),
+  READ_EXECUTE("r-x"),
+  READ_WRITE("rw-"),
+  ALL("rwx");
+
+  /** Symbolic representation */
+  public final String mSymbol;
```
Comment (apc999) [inline common/src/main/java/tachyon/security/authorization/FsAction.java]: we shall avoid exposing public member variable. Instead, providing a getter method to access this value.

File: common/src/main/java/tachyon/security/authorization/FsAction.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+      return (ordinal() & that.ordinal()) == that.ordinal();
+    }
+    return false;
+  }
+
+  /** AND operation. */
+  public FsAction and(FsAction that) {
+    return SVALS[ordinal() & that.ordinal()];
+  }
+  /** OR operation. */
```
Comment (apc999) [inline common/src/main/java/tachyon/security/authorization/FsAction.java]: NIT: better to leave a line between line 56 and 57

Comment (apc999) [inline common/src/main/java/tachyon/security/authorization/FsAction.java]: also applies to line 60 and 61

Comment (shenguoquan) [inline common/src/main/java/tachyon/security/authorization/FsAction.java]: Thanks for your remind. I will fix it

File: common/src/main/java/tachyon/Constants.java (hunk 1 above)
Comment (shenguoquan) [inline common/src/main/java/tachyon/Constants.java]: I will fix it.

File: common/src/main/java/tachyon/security/authorization/FsAction.java (hunk 2 above)
Comment (shenguoquan) [inline common/src/main/java/tachyon/security/authorization/FsAction.java]: I will change my code. Thanks

File: common/src/main/java/tachyon/Constants.java [hunk 4]
```diff
@@ -248,4 +248,7 @@
   public static final String TACHYON_SECURITY_AUTHENTICATION = "tachyon.security.authentication";
   public static final String TACHYON_AUTHENTICATION_PROVIDER_CUSTOM_CLASS =
       "tachyon.authentication.provider.custom.class";
+  // Authorization
+  public static final String FS_PERMISSIONS_UMASK_KEY = "tachyon.tfs.permission.umask";
```
Comment (ooq) [inline common/src/main/java/tachyon/Constants.java]: maybe just `TFS_PERMISSIONS_UMASK`?

File: common/src/main/java/tachyon/Constants.java [hunk 5]
```diff
@@ -248,4 +248,7 @@
...
+  public static final int DEFAULT_FS_PERMISSIONS_UMASK = 0022;
```
Comment (ooq) [inline common/src/main/java/tachyon/Constants.java]: `FS` -> `TFS`?

**Comment by ooq** [comment]
Minor comments left, more to follow.

File: common/src/main/java/tachyon/Constants.java (hunk 4 above)
Comment (shenguoquan) [inline common/src/main/java/tachyon/Constants.java]: I will rename the configure key. Thanks.

File: common/src/main/java/tachyon/Constants.java (hunk 5 above)
Comment (shenguoquan) [inline common/src/main/java/tachyon/Constants.java]: I will change it. THX

File: common/src/main/java/tachyon/security/authorization/FsPermission.java [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+  public FsPermission(short mode) {
+    fromShort(mode);
+  }
+
+  /**
+   * Copy constructor
+   *
+   * @param other other permission
+   */
+  public FsPermission(FsPermission other) {
```
Comment (ooq) [inline common/src/main/java/tachyon/security/authorization/FsPermission.java]: a little bit confusing: might mistake `other` here with otherAction

File: common/src/main/java/tachyon/security/authorization/FsPermission.java [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+  /** Apply a umask to this permission and return a new one */
+  public FsPermission applyUMask(FsPermission umask) {
+    return new FsPermission(mUseraction.and(umask.mUseraction.not()),
+        mGroupaction.and(umask.mGroupaction.not()),
+        mOtheraction.and(umask.mOtheraction.not()));
+  }
+
+  /** Get the default permission. */
+  public static FsPermission getDefault() {
+    return new FsPermission((short)0777);
```
Comment (ooq) [inline common/src/main/java/tachyon/security/authorization/FsPermission.java]: Should we make it configurable?

File: common/src/main/java/tachyon/security/authorization/FsPermission.java [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+   */
+  public static FsPermission getUMask(TachyonConf conf) {
+    int umask = Constants.DEFAULT_TFS_PERMISSIONS_UMASK;
+    if (conf != null) {
+      umask = conf.getInt(Constants.TFS_PERMISSIONS_UMASK_KEY,
+          Constants.DEFAULT_TFS_PERMISSIONS_UMASK);
+    }
+    return new FsPermission((short)umask);
+  }
+}
```
Comment (ooq) [inline common/src/main/java/tachyon/security/authorization/FsPermission.java]: Adding a newline would be nice.

File: common/src/test/java/tachyon/security/authorization/FsActionTest.java [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+    Assert.assertTrue(FsAction.ALL.implies(FsAction.READ));
+    Assert.assertTrue(FsAction.ALL.implies(FsAction.WRITE));
+    Assert.assertTrue(FsAction.ALL.implies(FsAction.EXECUTE));
+    Assert.assertTrue(FsAction.ALL.implies(FsAction.READ_EXECUTE));
+    Assert.assertTrue(FsAction.ALL.implies(FsAction.WRITE_EXECUTE));
+    Assert.assertTrue(FsAction.ALL.implies(FsAction.ALL));
+
+    Assert.assertTrue(FsAction.READ_EXECUTE.implies(FsAction.READ));
+    Assert.assertTrue(FsAction.READ_EXECUTE.implies(FsAction.EXECUTE));
+    Assert.assertFalse(FsAction.READ_EXECUTE.implies(FsAction.WRITE));
```
Comment (ooq) [inline common/src/test/java/tachyon/security/authorization/FsActionTest.java]: Complete this by checking `WRITE_EXECUTE` and `READ_WRITE`?

**Comment by ooq** [comment]
@shenguoquan Thanks for the PR. Overall it looks good! Just some minor comments.

File: common/src/main/java/tachyon/security/authorization/FsPermission.java (hunk 6 above)
Comment (shenguoquan) [inline common/src/main/java/tachyon/security/authorization/FsPermission.java]: Thanks. I will change it

File: common/src/main/java/tachyon/security/authorization/FsPermission.java (hunk 7 above)
Comment (shenguoquan) [inline common/src/main/java/tachyon/security/authorization/FsPermission.java]: yeah, Good advice. I will change it.

File: common/src/main/java/tachyon/security/authorization/FsPermission.java (hunk 8 above)
Comment (shenguoquan) [inline common/src/main/java/tachyon/security/authorization/FsPermission.java]: I will change it. Thanks

File: common/src/test/java/tachyon/security/authorization/FsActionTest.java (hunk 9 above)
Comment (shenguoquan) [inline common/src/test/java/tachyon/security/authorization/FsActionTest.java]: I will add it into unit test case. Thanks for your advice.

**Comment by shenguoquan** [comment]
Hi @ooq , I have updated my PR.

**Comment by ooq** [comment]
Is the newline added?

**Comment by shenguoquan** [comment]
Hi @ooq , I am sorry where I should add a newline? I can't find the place I should add. Please help me to point it out. Thanks.

**Comment by ooq** [comment]
@shenguoquan It's at the end of `FsPermission.java`.

**Comment by shenguoquan** [comment]
Hi @ooq , I have updated my PR according to your comments. Thanks

**Comment by ooq** [comment]
+1 LGTM

**Comment by apc999** [comment]
+1 LGTM

**Pull request merged**